rust/kernel/driver.rs

```diff
@@ -0,0 +1,205 @@
+// SPDX-License-Identifier: GPL-2.0
+
+//! Generic support for drivers of different buses (e.g., PCI, Platform, Amba, etc.).
+//!
+//! Each bus/subsystem is expected to implement [`DriverOps`], which allows drivers to register
+//! using the [`Registration`] class.
+
+use crate::{str::CStr, Error, KernelModule, Result, ScopeGuard, ThisModule};
+use alloc::{boxed::Box, vec::Vec};
+use core::{cell::UnsafeCell, mem::MaybeUninit, pin::Pin};
+
+/// A subsystem (e.g., PCI, Platform, Amba, etc.) that allows drivers to be written for it.
+pub trait DriverOps {
+    /// The type that holds information about the registration. This is typically a struct defined
+    /// by the C portion of the kernel.
+    type RegType: Default;
+
+    /// The type that holds identification data for the devices supported by the driver. In
+    /// addition to the information required by the bus, it may also store device-specific data
+    /// using Rust types.
+    type IdType: 'static;
```

Comment:
If you know `RegType` or `IdType` is there a single valid type for the other? If so maybe you can use a trait to get the type of one from the type of the other?

Reply:
Yes, we can define a trait that allows us to get `RegType` from `IdType`. I like the fact that we simplify this trait by eliminating a type.

However, to eliminate this type from this trait, we need to define yet another trait (with a similar type). It seems more complicated. Are there non-obvious advantages to what you're suggesting?

Reply:
I assumed that `DriverOps` would directly be used by the driver author, but in the next PR I see the subsystem specific code implements it, and the driver author needs to implement another trait specific for the subsystem.

Reply:
Ah, yes, I uploaded the usage of this module too to make it clearer to reviewers. Indeed this trait in only implemented by subsystems.